app/scripts/lib/channel-server-client.js

```diff
@@ -0,0 +1,294 @@
+
+        this.on('message', (event) => this._encryptedMessageHandler(event));
+        resolve();
+      };
+      /*eslint-enable no-use-before-define*/
+
+      const stopListeningToChannelSetupEvents = () => {
+        this.off('close', close);
+        this.off('error', error);
+        this.off('message', message);
+      };
+
+      this.on('close', close);
+      this.on('error', error);
+      this.once('message', message);
+    });
+  }
+
+  /**
+   * Close the channel.
+   *
+   * @returns {Promise} - rejects if no socket, resolves when connection closed
+   */
+  close () {
+    return new Promise((resolve, reject) => {
+      if (! this.socket) {
+        return reject(ChannelServerClientErrors.toError('NOT_CONNECTED'));
+      }
+
+      /*eslint-disable no-use-before-define*/
+      const stopListeningToChannelTeardownEvents = () => {
+        this.off('close', close);
+        this.off('error', error);
+      };
+
+      const close = () => {
+        stopListeningToChannelTeardownEvents();
+        this.set('isConnected', false);
+        resolve();
+      };
+
+      const error = (err) => {
+        stopListeningToChannelTeardownEvents();
+        reject(err);
+      };
+      /*eslint-enable no-use-before-define*/
+
+      this.on('close', close);
+      this.on('error', error);
+
+      this.socket.close();
+    });
+  }
+
+  /**
+   * Send `message` with `data` to the remote.
+   *
+   * @param {String} message name of message to send
+   * @param {Any} [data={}]
+   * @returns {Promise} resolves when complete
+   */
+  send (message, data = {}) {
+    if (! this.socket || ! this.get('isConnected')) {
+      return Promise.reject(ChannelServerClientErrors.toError('NOT_CONNECTED'));
+    }
+
+    const envelope = {
+      data,
+      message,
+    };
+
+    return this._encrypt(envelope)
+      .then(ciphertext => this.socket.send(ciphertext));
+  }
+
+  /**
+   * Proxy `socket` events as if they were from
+   * the client itself.
+   *
+   * @param {WebSocket} socket
+   * @private
+   */
+  _proxySocketEvents (socket) {
+    ['close', 'error', 'message', 'open'].forEach(eventName => {
+      socket.addEventListener(eventName, (event) => this.trigger(eventName, event));
+    });
+  }
+
+  /**
+   * Handle an encrypted message. Expects a message with the
+   * following format:
+   *
+   * ```
+   * {
+   *   "message": <encrypted message>,
+   *   "sender": {
+   *      "city": <sender's city name>,
+   *      "country": <sender's country name>,
+   *      "region": <sender's region>,
+   *      "ua": <sender's user agent string>
+   *   }
+   * }
+   * ```
+   *
+   * The encrypted message is expected to have the following format:
+   *
+   * ```
+   * {
+   *   "message": <message name>,
+   *   "data": <any, specific to the message>
+   * }
+   *
+   * @param {Object} event
+   * @private
+   */
+  _encryptedMessageHandler (event) {
+    const { message: ciphertext, sender } = JSON.parse(event.data);
```

Comment:
do we want to handle JSON.parse fails here? or just let it throw a generic parse error?

Reply:
We probably should throw a better error to easily find them in Sentry